Tasks/Maven/maventask.ts

```diff
@@ -3,97 +3,181 @@
+        mvnExec = tl.which('mvn', true);
+    }
+}
+tl.debug('Maven executable: ' + mvnExec);
 
+// Set JAVA_HOME to the JDK version (default, 1.7, 1.8, etc.) or the path specified by the user
+var specifiedJavaHome: string = null;
 if (javaHomeSelection == 'JDKVersion') {
-    tl.debug('Using JDK version to find and set JAVA_HOME');
-    var jdkVersion = tl.getInput('jdkVersion');
-    var jdkArchitecture = tl.getInput('jdkArchitecture');
+    // Set JAVA_HOME to the specified JDK version (default, 1.7, 1.8, etc.)
+    tl.debug('Using the specified JDK version to find and set JAVA_HOME');
+    var jdkVersion: string = tl.getInput('jdkVersion');
+    var jdkArchitecture: string = tl.getInput('jdkArchitecture');
 
     if (jdkVersion != 'default') {
         // jdkVersion should be in the form of 1.7, 1.8, or 1.10
         // jdkArchitecture is either x64 or x86
         // envName for version 1.7 and x64 would be "JAVA_HOME_7_X64"
-        var envName = "JAVA_HOME_" + jdkVersion.slice(2) + "_" + jdkArchitecture.toUpperCase();
+        var envName: string = "JAVA_HOME_" + jdkVersion.slice(2) + "_" + jdkArchitecture.toUpperCase();
         specifiedJavaHome = tl.getVariable(envName);
         if (!specifiedJavaHome) {
             tl.error('Failed to find specified JDK version. Please make sure environment variable ' + envName + ' exists and is set to the location of a corresponding JDK.');
             tl.exit(1);
         }
     }
-} else {
-    tl.debug('Using path from user input to set JAVA_HOME');
-    var jdkUserInputPath = tl.getPathInput('jdkUserInputPath', true, true);
+}
+else {
+    // Set JAVA_HOME to the path specified by the user
+    tl.debug('Setting JAVA_HOME to the path specified by user input');
+    var jdkUserInputPath: string = tl.getPathInput('jdkUserInputPath', true, true);
     specifiedJavaHome = jdkUserInputPath;
 }
 
+// Set JAVA_HOME as determined above (if different than default)
 if (specifiedJavaHome) {
     tl.debug('Set JAVA_HOME to ' + specifiedJavaHome);
     process.env['JAVA_HOME'] = specifiedJavaHome;
 }
 
-var publishJUnitResults = tl.getInput('publishJUnitResults');
-var testResultsFiles = tl.getInput('testResultsFiles', true);
+// Maven task orchestration occurs as follows:
+// 1. Check that Maven exists by executing it to retrieve its version.
+// 2. Run Maven with the user goals. Compilation or test errors will cause this to fail.
+// 3. Always try to run the SonarQube analysis if it is enabled.
+//    In case the build has failed, the analysis will still succeed but the report will have less data. 
+// 4. Always publish test results even if tests fail, causing this task to fail.
+// 5. If #2 above failed, exit with an error code to mark the entire step as failed. Same for #4.
 
-function publishTestResults(publishJUnitResults, testResultsFiles: string) {
-    var matchingTestResultsFiles: string[] = undefined;
+var userRunFailed: boolean = false;
```

Comment:
Why are we pulling the main function on top of other helper functions?  Is this the style now?